### PR TITLE
195 multiprojects

### LINKED ISSUE #195: Projektkonfiguration: Specmate konfiguriert alle Jira Projekte automatisch mit Default-Config
**User Story**
Als Administrator möchte ich, dass Specmate jira Projekte eines Jira-Servers automatisch konfiguriert, damit ich seltener ein Deployment durchführen muss.

**Acceptance Criteria**
- Specmate erlaubt die generische Konfiguration eines Jira-Connectors ohne die Angabe eines konkreten Projektnamens 
- Liegt eine solche generische Konfiguration vor, konfiguriert Specmate alles Jira Projekte, auf die der in der Konfiguration angegebene technische User Zugriff hat, mit den in den in der generischen Konfiguration angegebenen Daten
- Der Specmate Projektname setzt sich aus dem Namen der generischen Konfiguration und dem Namens des Jira Projekts
- Existiert für ein Jira Projekt jedoch eine spezifische Specmate Konfiguration nutzt Specmate diese Konfiguration und nicht die generische Konfiguration

### PR DESCRIPTION
**Pull request by hauptmab**
Closes #195

config example for testing:

```
# List of multi project ids, each project listed here needs to be configured below
multiproject.multiprojects=jiraLocalhost

# required
multiproject.jiraLocalhost.multiconnector.pid					= com.specmate.connectors.jira.JiraMultiConnector
multiproject.jiraLocalhost.multiconnector.jira.url				= http://127.0.0.1:2990/jira
multiproject.jiraLocalhost.multiconnector.jira.username 			= admin
multiproject.jiraLocalhost.multiconnector.jira.password 			= admin

#optional
multiproject.jiraLocalhost.projectnamepattern					= companyjira_%project%

# project template (depends on jira connector if these are required)
multiproject.jiraLocalhost.multiconnector.jira.template.withFolders		= false
multiproject.jiraLocalhost.multiconnector.jira.template.directJQL		= project=%project%

```

### CODE REVIEW

**Comment by junkerm** [comment]
if no projects (project.projects) are configured I get a NullPointerException: 
grafik

**Comment by junkerm** [comment]
If i configure the jira multiconnector and a regular jira connector (for the same jira isntance) none of the multiprojects are shown:  
grafik

**Comment by junkerm** [comment]
Is it already possible to configure libraries (e.g. via template)?

File: bundles/specmate-connectors-api/src/com/specmate/connectors/api/IMultiConnector.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+	/**
+	 * Returns the id of the connector
+	 */
+	String getId();
+
+	/**
+	 * Returns the project the connector is associated with
+	 * 
+	 * @return
```
Comment (junkerm) [inline bundles/specmate-connectors-api/src/com/specmate/connectors/api/IMultiConnector.java]: remove empty @return

File: bundles/specmate-connectors-api/src/com/specmate/connectors/api/IMultiConnector.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package com.specmate.connectors.api;
+
+import java.util.Map;
+
+import com.specmate.common.exception.SpecmateException;
+
+public interface IMultiConnector {
```
Comment (junkerm) [inline bundles/specmate-connectors-api/src/com/specmate/connectors/api/IMultiConnector.java:13]: Please add short interface documentation

File: bundles/specmate-connectors-api/src/com/specmate/connectors/api/IMultiConnector.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	 */
+	IMultiProject getMultiProject();
+
+	/**
+	 * Associcates the connector with a project
+	 */
+	void setMultiProject(IMultiProject multiProject);
+
+	/**
+	 * Returns a list of project configs. Key is project id as it is known in the
```
Comment (junkerm) [inline bundles/specmate-connectors-api/src/com/specmate/connectors/api/IMultiConnector.java]: list -> map

**Review by junkerm — COMMENTED** [review]
when i configure two multiproject for the same jira instance i get a nullpointer exception
grafik

File: bundles/specmate-connectors/src/com/specmate/connectors/internal/MultiConnectorService.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+import com.specmate.connectors.api.IMultiConnector;
+import com.specmate.connectors.api.IMultiProject;
+import com.specmate.connectors.api.IProjectConfigService;
+import com.specmate.connectors.internal.config.MultiConnectorServiceConfig;
+import com.specmate.connectors.internal.config.PollKeys;
+import com.specmate.scheduler.Scheduler;
+import com.specmate.scheduler.SchedulerIteratorFactory;
+import com.specmate.scheduler.SchedulerTask;
+
+@Component(immediate = true, configurationPid = MultiConnectorServiceConfig.PID, configurationPolicy = ConfigurationPolicy.REQUIRE)
```
Comment (junkerm) [inline bundles/specmate-connectors/src/com/specmate/connectors/internal/MultiConnectorService.java:32]: please add a short documentation

File: bundles/specmate-connectors/src/com/specmate/connectors/internal/MultiConnectorService.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+		if (schedule == null) {
+			logService.log(LogService.LOG_INFO, "Polling interval '" + PollKeys.KEY_POLL_SCHEDULE + "' not set.");
+			return;
+		}
+
+		MultiConnectorTask task = new MultiConnectorTask(this, logService);
+
+		Scheduler scheduler = new Scheduler();
+		scheduler.schedule(task, SchedulerIteratorFactory.create(schedule));
+	}
```
Comment (junkerm) [inline bundles/specmate-connectors/src/com/specmate/connectors/internal/MultiConnectorService.java:60]: you could add a @Deactivate method where you stop the schedular

File: bundles/specmate-connectors/src/com/specmate/connectors/internal/MultiConnectorService.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+		private MultiConnectorService multiConnectorService;
+		private LogService logService;
+
+		public MultiConnectorTask(MultiConnectorService multiConnectorService, LogService logService) {
+			this.multiConnectorService = multiConnectorService;
+			this.logService = logService;
+		}
+
+		@Override
+		public void run() {
```
Comment (junkerm) [inline bundles/specmate-connectors/src/com/specmate/connectors/internal/MultiConnectorService.java:86]: format (or at leas remove empty lines)

File: bundles/specmate-connectors/src/com/specmate/connectors/internal/MultiConnectorService.java [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+			List<String> configuredProjects = Arrays
```
Comment (junkerm) [inline bundles/specmate-connectors/src/com/specmate/connectors/internal/MultiConnectorService.java]: would it make sense to retrieve the configured projects from the project service? What if a different multiconnectorservice is adding projects in parallel?

File: bundles/specmate-connectors/src/com/specmate/connectors/internal/MultiProjectImpl.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+import com.specmate.common.exception.SpecmateInternalException;
+import com.specmate.connectors.api.IMultiConnector;
+import com.specmate.connectors.api.IMultiProject;
+import com.specmate.connectors.api.IProjectConfigService;
+import com.specmate.connectors.config.ProjectConfigService;
+import com.specmate.model.administration.ErrorCode;
+
+@Component(immediate = true, service = IMultiProject.class, configurationPid = ProjectConfigService.MULTIPROJECT_CONFIG_FACTORY_PID, configurationPolicy = ConfigurationPolicy.REQUIRE)
+public class MultiProjectImpl implements IMultiProject {
```
Comment (junkerm) [inline bundles/specmate-connectors/src/com/specmate/connectors/internal/MultiProjectImpl.java:23]: please add short documentatino

File: bundles/specmate-connectors/src/com/specmate/connectors/internal/config/MultiConnectorServiceConfig.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.log.LogService;
+
+import com.specmate.common.OSGiUtil;
+import com.specmate.common.exception.SpecmateException;
+import com.specmate.config.api.IConfigService;
+
+@Component(immediate = true)
+public class MultiConnectorServiceConfig {
```
Comment (junkerm) [inline bundles/specmate-connectors/src/com/specmate/connectors/internal/config/MultiConnectorServiceConfig.java:20]: please add short documenation

File: bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraMultiConnector.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+	/**
+	 * Crates a list of all template properties (without prefix).
+	 */
+	private static Map<String, String> getTemplateProperties(Map<String, Object> properties) {
+		Map<String, String> templateProperties = new HashMap<>();
+		for (String propertyKey : properties.keySet()) {
+			if (propertyKey.startsWith(JiraConfigConstants.KEY_JIRA_MULTIPROJECT_TEMPLATE_PREFIX)) {
+				templateProperties.put(
+						propertyKey.substring(JiraConfigConstants.KEY_JIRA_MULTIPROJECT_TEMPLATE_PREFIX.length()),
```
Comment (junkerm) [inline bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraMultiConnector.java]: you could pull out the length() as variable at the start of the method

**Review by junkerm — CHANGES_REQUESTED** [review]
there are some small issues when no regular project is configured and when several multiprojects are configured (see comments) otherwiese just minor things

**Comment by hauptmab** [comment]
> if no projects (project.projects) are configured I get a NullPointerException:
> grafik

fixed

**Comment by hauptmab** [comment]
> If i configure the jira multiconnector and a regular jira connector (for the same jira isntance) none of the multiprojects are shown:
> grafik

@junkerm Is this the behaviour we talked about yesterday? Is it still not working?

**Comment by hauptmab** [comment]
> Is it already possible to configure libraries (e.g. via template)?

this is handled in another issue: https://github.com/qualicen/specmate/issues/255

**Comment by hauptmab** [comment]
> when i configure two multiproject for the same jira instance i get a nullpointer exception
> grafik

This definitely was a bug. Fix already pushed.

File: bundles/specmate-connectors/src/com/specmate/connectors/internal/MultiConnectorService.java (hunk 7 above)
Comment (hauptmab) [reply bundles/specmate-connectors/src/com/specmate/connectors/internal/MultiConnectorService.java]: Good point. However, during startup, it is not garanteed that all projects have already been created when the multi connectors will run the first time. Looking at the config will uncover projects which are not actived yet.

**Comment by junkerm** [comment]
> 
> 
> > If i configure the jira multiconnector and a regular jira connector (for the same jira isntance) none of the multiprojects are shown:
> > grafik
> 
> @junkerm Is this the behaviour we talked about yesterday? Is it still not working?

it works now

**Review by junkerm — APPROVED** [review]

**Pull request merged**